### PR TITLE
WA-338 Pairing the browser extension and mobile apps

### PR DESCRIPTION
**Pull request by germartinez**
It allows to pair the browser extension and the mobiles apps using Firebase and the push notification service.

In order to do the pairing between the browser extension and the mobile app, a pushToken is requested to Firebase and sent to the server.

There is a new required permission in the manifest: "notifications".

The PairingProcess component has been refactored into 3 smaller ones:
- CreateAccount
- LockedAccount
- UnlockedAccount

### CODE REVIEW

File: app/routes/PairingProcess/components/CreateAccount/index.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    const mnemonic = Bip39.generateMnemonic()
+    const {
+      currentAccount,
+      encryptedMnemonic,
+      hmac,
+    } = createEthAccount(mnemonic, password)
+
+    setUpNotifications()
+      .then((token) => {
+        if (token !== null) {
```
Comment (apanizo) [inline app/routes/PairingProcess/components/CreateAccount/index.js]: Could be simplified:
```
if (token === null) {
  ...
  return  
}
if (authPushNotificationService(token, currentAccount.getPrivateKey())) {
    ...
}
```

File: app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    'Content-Type': 'application/json',
+  }
+  const body = generateAuthContent(pushToken, privateKey)
+
+  return fetch(url, {
+    method: 'POST',
+    headers,
+    body,
+  })
+    .then((response) => {
```
Comment (apanizo) [inline app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js]: ```.then( response => response.status === 201)```
The question is, as far as I remember if you receive other status different than 2XX the ```catch``` will be executed.

Why are you just checking 201? What about the others 20X?

File: app/routes/PairingProcess/components/CreateAccount/index.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          if (authPushNotificationService(token, currentAccount.getPrivateKey())) {
+            this.setState({ loading: false })
+            this.props.onCreateAccount(currentAccount.getChecksumAddressString(), encryptedMnemonic, hmac)
+          }
+        }
+        else {
+          this.setState({
+            loading: false,
+            message: 'Permission to receive notifications required.'
```
Comment (apanizo) [inline app/routes/PairingProcess/components/CreateAccount/index.js]: Are you sure that is the only case for failing?

File: app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        ? true
+        : false
+    })
+    .catch((err) => {
+      console.error(err)
+      return false
+    })
+}
+
+const generateAuthContent = (pushToken, privateKey) => {
```
Comment (apanizo) [inline app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js]: Just for my interest :), can you explain to me what is going on here?

File: app/routes/PairingProcess/components/Layout.js [hunk 5]
```diff
@@ -1,13 +1,20 @@
 import React, { Component } from 'react'
```
Comment (apanizo) [inline app/routes/PairingProcess/components/Layout.js]: NOTE TO ME: Reviewed until here

File: app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js (hunk 2 above)
Comment (germartinez) [reply app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js]: The push notification service specifications indicate that a HTTP 201 is returned if OK

File: app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js (hunk 4 above)
Comment (germartinez) [reply app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js]: This message is sent to the push notification service from the extension in order to be authorized to receive notifications and to pair the extension with the mobile app.
```
{
  "pushToken": <string>,
  "signature": { // signs sha3("GNO" + <pushToken>)
    "v": <integer>,
    "r": <string>, # stringified int
    "s": <string> # stringified int
  },
}
```

File: app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js (hunk 2 above)
Comment (apanizo) [reply app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js]: I think if you just do:
```
return fetch(url, {
    method: 'POST',
    headers,
    body,
 })
```
And in the class where you use this function:
```
try {
    await authPushNotificationService(token, currentAccount.getPrivateKey())    
} catch (err) {
    console.error(err)
}
this.setState({ loading: false })
this.props.onCreateAccount(currentAccount.getChecksumAddressString(), encryptedMnemonic, hmac)
```
You achieve the same in a cleaner way. But this is just my opinion. ;)

File: app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js (hunk 4 above)
Comment (apanizo) [reply app/routes/PairingProcess/components/CreateAccount/pairingNotifications.js]: Got it. Really clever the push notification service. 

Anyway, can you leave a comment just before the function with the explanation you have just described? Probably in a couple of weeks I will forget it :)

File: app/routes/PairingProcess/components/Layout.js (hunk 5 above)
Comment (apanizo) [reply app/routes/PairingProcess/components/Layout.js]: @germartinez Let me know when comments are addressed until here.

File: app/routes/PairingProcess/containers/PairingProcess.test.js [hunk 6]
```diff
@@ -69,19 +75,19 @@ describe('PairingProcess', () => {
 
     const component = shallow(
       <PairingProcess location={location} store={mockStore} />
-    ).dive()
+    ).dive().dive()
 
-    expect(mockStore.getActions().length).toBe(0)
+    expect(component.find('Connect(LockedAccount)')).toHaveLength(1)
```
Comment (apanizo) [inline app/routes/PairingProcess/containers/PairingProcess.test.js]: What is ```'Connect(xxx)'```?

**Pull request merged**